darling/src/libsystem_kernel/emulation/src/xnu_syscall/bsd/impl/misc/clonefile.c

```diff
@@ -1,13 +1,224 @@
+#include <darling/emulation/xnu_syscall/bsd/impl/fcntl/open.h>
+#include <darling/emulation/xnu_syscall/bsd/impl/fcntl/openat.h>
+#include <darling/emulation/xnu_syscall/bsd/impl/stat/mkdirat.h>
+#include <darling/emulation/xnu_syscall/bsd/impl/unistd/close.h>
+#include <darling/emulation/xnu_syscall/bsd/impl/unistd/fchmod.h>
+#include <darling/emulation/xnu_syscall/bsd/impl/unistd/fchown.h>
+#include <darling/emulation/xnu_syscall/bsd/impl/unistd/unlink.h>
+
+#ifndef MAX_COPY_CHUNK
+#define MAX_COPY_CHUNK ((size_t)0x7ffff000)
+#endif
 
 long sys_clonefileat(int src_fd, const char* src_path, int dest_fd, const char* dest_path, uint32_t flags) {
 	return -ENOSYS;
 }
 
 long sys_fclonefileat(int src_fd, int dest_fd, const char* dest_path, uint32_t flags) {
-	return -ENOSYS;
+    struct stat st;
+
+    if (fstat(src_fd, &st) < 0) {
+        return -errno;
+    }
+
+    if (!S_ISREG(st.st_mode)) {
+        return -EOPNOTSUPP;
+    }
+
+    if (flags & ~CLONE_NOFOLLOW) {
+        return -EINVAL;
+    }
+
+    int open_flags = BSD_O_CREAT | BSD_O_WRONLY | BSD_O_EXCL;
+
+    if (flags & CLONE_NOFOLLOW) {
+        open_flags |= BSD_O_NOFOLLOW;
+    }
+
+    if (
+        (flags & CLONE_NOFOLLOW_ANY)
+        || (flags & CLONE_ACL)
+        || (flags & CLONE_RESOLVE_BENEATH)
+    ) {
+        return -ENOTSUP;
+    }
+
+    int saved_errno = 0;
+    int ret = 0;
+    char *filename;
+    
+    if (dest_fd == BSD_AT_FDCWD) {
+        char *path = strdup(dest_path);
+
+        if (path[0] == '/') {
+            dest_fd = sys_open("/", BSD_O_RDONLY | BSD_O_DIRECTORY | BSD_O_CLOEXEC, 0);
+            path++;
+        } else {
+            dest_fd = LINUX_AT_FDCWD;
+        }
+
+        char *token;
+        int has_error = 0;
+
+        while ((token = strsep(&path, "/")) != NULL) {
+            if (*token == '\0') continue;
+
+            int next_fd = sys_openat(dest_fd, token, BSD_O_RDONLY | BSD_O_DIRECTORY | BSD_O_CLOEXEC, st.st_mode & 0777);
+
+            if (next_fd >= 0) {
+                if (dest_fd != LINUX_AT_FDCWD) sys_close(dest_fd);
+
+                dest_fd = next_fd;
+                continue;
+            }
+
+            if (path == NULL) {
+                filename = token;
+                break;
+            }
+
+            if (errno != ENOENT) {
+                saved_errno = errno;
+                ret = -saved_errno;
+                goto out_close;
+            };
+
+            if (sys_mkdirat(dest_fd, token, 0755) < 0) {
+                saved_errno = errno;
+                ret = -saved_errno;
+                goto out_close;
+            }
+
+            next_fd = sys_openat(dest_fd, token, BSD_O_RDONLY | BSD_O_DIRECTORY | BSD_O_CLOEXEC, st.st_mode & 0777);
+
+            if (next_fd < 0) {
+                saved_errno = errno;
+                ret = -saved_errno;
+                goto out_close;
+            }
+
+            if (dest_fd != LINUX_AT_FDCWD) sys_close(dest_fd);
+
+            dest_fd = next_fd;
+        }
+    }
```

Comment:
Help me better understand the big picture of this block
* Why did you design the inside the block to only execute when `dest_fd == BSD_AT_FDCWD`?